### ISSUE TITLE
Create lib.common.d.ts?

### ISSUE THREAD

**Issue by saschanaz** [microsoft/TypeScript-DOM-lib-generator#1402]
Situations like https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1368 will keep happening as the commonly implemented interfaces update their definition time to time. It might be worth to have a separate library for this.

The requirement to enter the common lib would be 2+ browser engine implementations (Gecko, Blink, WebKit) + 2+ non-browser implementations (node, deno, etc.)

**Comment by jimmywarting** [comment]
I stumble upon another issue today that 

## TypeScript thinks that NodeJS web streams are not the same as DOM web streams

When i tried this code it throws an error saying that they are not compatible with each other.
That they are not exactly matching.

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/1148376/96863da2-3a7f-42f8-9fc4-d6f304df0152)

one comes from `@types/node` and the other one comes from `lib.dom`
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/1148376/b655b0cf-ca59-4b31-bb4e-a40b6131c796)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/1148376/ff3c73af-c215-48ef-8d83-4f56aa755ca1)


![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/1148376/edaf35d7-b3d3-4b95-b7e6-9a91d25cc3da)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/1148376/2c439477-51fa-432a-b5ce-cadc88bf23fd)

Even tough they are exactly the same `globalThis.ReadableStream === require('stream/web').ReadableStream`

You should definitely consider splitting up `lib.dom` into one that every environment have built in in both browser and non-browser context (that also includes different web worker).

> + "2+ non-browser implementations (node, deno, etc.)".
think Bun.js should be part of this

I think that that `common.d.ts` should should just be defined by the [wintercg common min api](https://common-min-api.proposal.wintercg.org/#index) rather than defining some kind of minimum rule that at least 2 should have it.

just the other week i had a beef with atob / btoa where it said that they where deprecated in NodeJS but not the DOM

So i think that it would be worth it if every project extended `lib.common.d.ts`

**Comment by jimmywarting** [comment]
and btw, i think all the issues in @saschanaz https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1402#issue-1377066431 where he linked to a few issues should be closed as a duplicate of this issue as i think this is the right path forward.

> closing as dupl of [microsoft/TypeScript-DOM-lib-generator#1402](microsoft/TypeScript-DOM-lib-generator#1402)

**Comment by saschanaz** [comment]
> I think that that `common.d.ts` should should just be defined by the [wintercg common min api](https://common-min-api.proposal.wintercg.org/#index) rather than defining some kind of minimum rule that at least 2 should have it.

That's in no way official at this point, so I'd rather not use it.


> where he linked to

Wrong. 😉

**Comment by hax** [comment]
I meet a similar issue. You can't use `for await` on `ReadableStream` because `AsyncIterable` interface has not been added into `dom` lib or `@types/web`. Even you use `@types/node`, if you import `ReadableStream` from `node:stream/web`, you can use `for await`, but the global `ReadableStream` doesn't work though they actually are same in node env. There might be other similar issues. It's clear that we need `lib.common.d.ts`.

And we also want `@types/wintercg-api` or some similar for the use cases which the user want to ensure the code only use common APIs crossed web, node, deno, bun, cloudflare workers and so on.

**Comment by Bashamega** [comment]
Hello @saschanaz,
Can I try to work on this issue?
I will use BCD to create the common lib, and add a common boolean to the JSON files.

**Comment by saschanaz** [comment]
> I will use BCD to create the common lib

This should be sufficient.

> add a common boolean to the JSON files.

With BCD this should not be needed.

I wonder we should do #222 first, in case implementations have their own extensions on the class objects.

**Comment by jakebailey** [comment]
I don't think this is a trivial thing unless you're meaning to just have it be part of the build process. TypeScript has to know the exact names of all of these libs, so we can't add anything new here and not also change things in TS, which needs to be driven from a feature request in TS itself since it needs to be a valid lib entry.

**Comment by saschanaz** [comment]
I was mostly thinking of publishing as `@types` package, which is being done independent of TypeScript itself.
